### ISSUE TITLE
make-vocab on multiple name fields

### ISSUE THREAD

**Issue by junkyul** [junkyul/allenel#2]
name fields:
mention_surface: mention words shown in a sentence (context)
mention_sentence: a sentence 
coherehce_labels: mention_surfaces shown together in a wiki page
type_labels: typically 1 or 2 of the 113 FIGER types for the mention_surface

- DatasetReader reads training examples and returns Instance(fields) per example

- Instance is a Map of fields with the key given inside the reader 
(Instance is created by passing a dict of fields)
each field is some container (TextField, LabelField, MetaField, etc)

- Vocabulary is also a mapping of dicts between text and index, where
each namespace (key of a field) is the key to each field.
There is only 1 Vocabulary object in the program.

- once it is created for Fields supporting vocabularies (Almost all except Meta?)
Vocabulary object should contain vocabularies for 
mention_surface, mention_sentence, coherehce_labels, type_labels, etc

$ allennlp make-vocab -h
usage: allennlp make-vocab [-h] -s SERIALIZATION_DIR [-o OVERRIDES]
                           [--include-package INCLUDE_PACKAGE]
                           param_path

Create a vocabulary from the specified dataset.

positional arguments:
  param_path            path to parameter file describing the model and its inputs

optional arguments:
  -h, --help            show this help message and exit
  -s SERIALIZATION_DIR, --serialization-dir SERIALIZATION_DIR
                        directory in which to save the vocabulary directory
  -o OVERRIDES, --overrides OVERRIDES
                        a JSON structure used to override the experiment
                        configuration
  --include-package INCLUDE_PACKAGE
                        additional packages to include

**Comment by junkyul** [comment]
- GloVe embeddings 300 dim
    training_config/naqanet.jsonnet
    https://s3-us-west-2.amazonaws.com/allennlp/datasets/glove/glove.840B.300d.lower.converted.zip

    training_config/biattentive_classification_network.jsonnet
    training_config/decomposable_attention.jsonnet
    training_config/bimpm.jsonnet
    training_config/coref.jsonnet
    training_config/biattentive_classification_network_elmo.jsonnet
    https://s3-us-west-2.amazonaws.com/allennlp/datasets/glove/glove.840B.300d.txt.gz

- What  is start and end tokens for a sentence, what is used in GloVe embedding?
    from allennlp.common.util import START_SYMBOL, END_SYMBOL 
    seq2seq.py
    https://github.com/allenai/allennlp/issues/2656

- embedding dropout instead of word dropout
    biattentive_classification_network_elmo.jsonnet
    "embedding_dropout": 0.5,

**Comment by junkyul** [comment]
load vocab from a directory

```
"vocabulary": {
directory_path: PATH,
pretrained_files: Optional[Dict[str, str]] = None,
only_include_pretrained_words: bool = False,
tokens_to_add: Dict[str, List[str]] = None,
min_pretrained_embeddings: Dict[str, int] = None   
}
```
from_params -> from_files -> from_instances

How to use all the options?
passing Dict[str, List[str]] ??? by https://jsonnet.org/

**Comment by junkyul** [comment]
current make_vocab reads all data at once (also for the lazy mode).
when text data is ~ 28 GB, it does not finish making vocab due to memory out (64 GB machine).
no way but fixing library code?
use machine with more memory... (it's one time effort)

**Comment by junkyul** [comment]
make vocab with one file, and extend it
```
  "vocabulary": {
    "directory_path": PATH_TO_VOCAB,
    "extend" : true
  },
```
1 training file has  10^6 examples
< 2 min reading, < 1 min fitting
12 GB memory
total < 5 min

**Comment by junkyul** [comment]
vocab files are just enumeration of words;
easy to create them from outside of allennlp. 
when extending, make_vocab should not allow duplicated tokens.

**Comment by junkyul** [comment]
MultiLabel creates One Hot Vector.
For coherence labels, it's creating |Coherence| Vector
For candidate labels, it's also creating |WID| Vector --> avoid this!

TextField can have variable length batches (no one hot encoding for all basis)

Also, see separate embedders for multiple inputs
https://github.com/allenai/allennlp/issues/2333

**Comment by junkyul** [comment]
To use multiple TextFields with different vocabs,
```
{
  "dataset_reader": {
    "type": "seq2seq",
    "source_tokenizer": {
      "type": "word",
      "word_splitter": {
        "type": "spacy",
        "pos_tags": true,
        "parse": true,
        "ner": true
      }
    },
    "target_tokenizer": {
      "type": "word"
    },
    "source_token_indexers": {
      "tokens": {
        "type": "single_id",
        "namespace": "source_tokens"
      },
      "pos_tags": {
        "type": "pos_tag",
        "namespace": "pos"
      },
      "dependency_label": {
        "type": "dependency_label",
        "namespace": "dependencies"
      },
      "ner_tags": {
        "type": "ner_tag",
        "namespace": "ner"
      }
    },
    "target_token_indexers": {
      "tokens": {
        "namespace": "target_tokens"
      }
    }
  },
```

change reader and model using TextFields, and Embeddings.
(what is the right config for Embedding, TextEmbedding, TokenEmbedding?)

current token vocab contains words out of public GloVe.
extend word embedding? best if one can set weights directly;
or make glove from the code and the data
or train random embedding.. (takes longer time to train?)

json using multiple namespace textfields
https://github.com/allenai/allennlp/blob/master/allennlp/tests/fixtures/encoder_decoder/simple_seq2seq/experiment.json
seq2seq reader 
https://github.com/allenai/allennlp/blob/master/allennlp/data/dataset_readers/seq2seq.py
simple seq2seq model https://github.com/allenai/allennlp/blob/master/allennlp/models/encoder_decoders/simple_seq2seq.py

**Comment by junkyul** [comment]
```
  "source_token_indexers": {
      "tokens": {
        "type": "single_id",
        "namespace": "source_tokens"
      },
      "pos_tags": {
        "type": "pos_tag",
        "namespace": "pos"
      },
   }
```
Indexers see tokens (text), tags, or labels; tages and labels are unpadded vocabs.
TextField creates a dictionary of tensors from input data;
indexer creates indexes for each key (namespace? type?) of data;

TextEmbedder reads this dictionary and embeds its subtypes/namespaces and create a tensor enclosing them all at once?; "token" is not for vocabulary namespace here.

**Comment by raman-r-4978** [comment]
Can any of you help me with proper examples of **make_vocab** command and its usages during **train** command?

**Comment by raman-r-4978** [comment]
I have the following json

> {
  "train_data_path": "data/train.csv",
  "validation_data_path": "data/validation.csv",
  "vocabulary": {
    "min_count": {
      "tokens": 10
    }
  },
  "dataset_reader": {
    "type": "classification_data_reader",
    "max_seq_len": 100,
    "tokenizer": {
      "type": "word",
      "word_splitter": {
        "type": "bert-basic",
        "do_lower_case": true
      }
    },
    "token_indexers": {
      "namespace": "tokens",
      "bert": {
        "type": "bert-pretrained",
        "pretrained_model": "bert-base-uncased",
        "do_lowercase": true
      }
    }
  }
}

when I run **allennlp make-vocab experiments/config.json -s vocab_dir --include-package library**

i) The commands runs fine.
ii) It creates a directory **vocab_dir** and inside it has **vocabulary** directory which has labels.txt and non_padded_namespaces.txt 

i.e.
-  vocab_dir
    - vocabulary
        - labels.txt (contains my binary labels 0 and 1)
        - non_padded_namespaces.txt (contains *tags and *labels)

other than this, I don't see vocabulary directory contains any of the information realted to vocabulary tokens. 
- What I am asking is where do you store my dataset tokens?
- Is this expected behavior? if not please help me with this.
